File: 0001.py
"""
    Use hash map which built-in in Python as dictionary
    key: numbers, value: index
    T: O(n)/S: O(n)
"""
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        x = {}
        for i, num in enumerate(nums):
            n = target - num
            if n in x:
                return [x[n], i]
            else:
                x[num] = i

File: 0001.cpp
// Second time
/*
    hash map (key: num, value: index)
    Do search and store at 1 pass
    *Note: Because solution is unique, num won't be duplicated -> can be key of map
    T: O(n)/S: O(n)
*/
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> mp;
        for (int i=0; i<nums.size(); ++i) {
            if (mp.count(target-nums[i]))
                return {mp[target-nums[i]], i};
            mp[nums[i]]=i;
        }
        return {};
    }
};


// First time
/*
    Use hash map
    key: number, value: index
    T: O(n)/S:(n)
*/
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int> mp;
        for (int i = 0; i < nums.size(); ++i) {
            if (mp.find(target-nums[i]) != mp.end())
                return {mp[target-nums[i]], i};
            mp[nums[i]] = i;
        }
        return {};
    }
};